patches/tModLoader/Terraria/ModLoader/BossBarLoader.cs

```diff
@@ -0,0 +1,323 @@
+		/// </summary>
+		/// <param name="bossBar">The ModBossBar</param>
+		/// <returns>Its texture, or the vanilla texture</returns>
+		public static Asset<Texture2D> GetTexture(ModBossBar bossBar) {
+			if (!bossBarTextures.TryGetValue(bossBar.index, out Asset<Texture2D> texture))
+				texture = VanillaBossBarTexture;
+			return texture;
+		}
+
+		/// <summary>
+		/// Gets the ModBossBar associated with this NPC
+		/// </summary>
+		/// <param name="npcType">The NPC type</param>
+		/// <param name="value">When this method returns, contains the ModBossBar associated with the specified NPC type</param>
+		/// <returns><see langword="true"/> if ModBossBar exists; otherwise, <see langword="false"/>.</returns>
+		public static bool NpcToBossBar(int npcType, out ModBossBar value) {
+			value = null;
+			if (npcToBossBarType.TryGetValue(npcType, out int index))
+				value = bossBars[index];
+			return value != null;
+		}
+
+		/// <summary>
+		/// Matches NPC types to their ModBossBar
+		/// </summary>
+		internal static void BuildReverseLookup() {
+			for (int i = 0; i < bossBars.Count; i++) {
+				ModBossBar bossBar = bossBars[i];
+				var validNPCs = bossBar.ValidNPCs;
+
+				//The same NPC can exist in multiple bars. In that case, only take the first found occurence
+				foreach (int npcType in validNPCs) {
+					if (!npcToBossBarType.ContainsKey(npcType))
+						npcToBossBarType.Add(npcType, i);
+				}
+			}
+		}
+
+		/// <summary>
+		/// Inserts the boss bar style select option into the main and ingame menu under the "Interface" category
+		/// </summary>
+		internal static string InsertMenu(out Action onClick) {
+			string styleText = null;
+			ModBossBarStyle pendingBossBarStyle = null;
+			foreach (ModBossBarStyle bossBarStyle in bossBarStyles) {
+				if (bossBarStyle == CurrentStyle) {
+					styleText = bossBarStyle.DisplayName;
+					break;
+				}
+
+				pendingBossBarStyle = bossBarStyle;
+			}
+
+			if (pendingBossBarStyle == null)
+				pendingBossBarStyle = bossBarStyles.Last();
+
+			if (styleText == null || bossBarStyles.Count == 1)
+				styleText = Language.GetTextValue("tModLoader.BossBarStyleNoOptions");
+
+			onClick = () => SwitchBossBarStyle(pendingBossBarStyle);
+
+			return Language.GetTextValue("tModLoader.BossBarStyle", styleText);
+		}
+
+		public static bool PreDraw(SpriteBatch spriteBatch, BigProgressBarInfo info, ref BossBarDrawParams drawParams) {
+			int index = info.npcIndexToAimAt;
+			if (index < 0 || index > Main.maxNPCs)
+				return false; //Invalid data, abort
+
+			NPC npc = Main.npc[index];
+
+			bool modify = true;
+
+			if (NpcToBossBar(npc.type, out ModBossBar bossBar)) {
+				drawParams.barTexture = GetTexture(bossBar).Value;
+				modify = bossBar.PreDraw(spriteBatch, npc, ref drawParams);
+			}
+
+			foreach(GlobalBossBar globalBossBar in globalBossBars) {
+				modify &= globalBossBar.PreDraw(spriteBatch, npc, ref drawParams);
+			}
```

Comment:
Would be worth rearranging this so GlobalBossBar is checked first, and only call ModBossBar if the global predraw returned true.

See implementation of mixed mod/global as seen in https://github.com/tModLoader/tModLoader/blob/5e311a223dc23b0ded123c5a0ae13c880528f477/patches/tModLoader/Terraria.ModLoader/NPCLoader.cs#L698

Reply:
Yes, globals should be able to have full control of ModX, and be able to prevent its hooks from running in cases like this one.